File: frontend/src/api/cvDrafts.ts
import { useQuery } from "@tanstack/react-query";

import { apiFetch } from "./client";
import type { CVDraft, CVDraftSummary } from "./models";

// Phase 20 — create/update/remove reuse useEntityMutations<CVDraft>
// (api/entities.ts) directly against pathSegment "drafts" — CVDraft is
// registered in api/routes/entity_crud.py's generic factory like every
// other entity, so there's nothing new to wrap for those three. Only
// list/get-one need their own hooks: they're hand-written on the backend
// (api/routes/cv_drafts.py, mirroring api/evidence.ts's own precedent —
// the factory has never had a GET verb for any entity).

export function useCvDrafts(candidateId: string | null) {
  return useQuery({
    queryKey: ["candidates", candidateId, "drafts"],
    queryFn: () => apiFetch<CVDraftSummary[]>(`/candidates/${candidateId}/drafts`),
    enabled: candidateId != null,
  });
}

// `enabled` (default true) lets a caller stop observing this query
// without unmounting — DraftScreen.tsx passes `false` once its own
// delete-draft mutation starts, so this actively-mounted query can't
// still be invalidated-and-refetched (a real, reproducible race: the
// remove mutation's own cache invalidation was refetching this exact
// draft — already gone server-side — for the split second before
// `navigate()` away could unmount the component and stop it, and that
// concurrent refetch was intermittently aborting the DELETE request
// itself, net::ERR_ABORTED, leaving the UI stuck on "Deleting…" even
// though the delete had already succeeded).
export function useCvDraft(candidateId: string | null, draftId: string | null, enabled = true) {
  return useQuery({
    queryKey: ["candidates", candidateId, "drafts", draftId],
    queryFn: () => apiFetch<CVDraft>(`/candidates/${candidateId}/drafts/${draftId}`),
    enabled: enabled && candidateId != null && draftId != null,
  });
}

File: api/routes/cv_drafts.py
"""GET /candidates/{candidate_id}/drafts (list) + GET .../drafts/{draft_id}
(one, full) — Phase 20's CVDraft entity has no update/create/remove code
here at all; those go through api/routes/entity_crud.py's generic factory
(registered in `_ENTITY_REGISTRATIONS`) exactly like every entity since
Phase 15a. This module exists only because the factory has no GET verb
for any entity — same reasoning as candidates.py's hand-written bulk
Evidence GET.

`list_cv_drafts` (below) is deliberately not the raw CandidateService
method name pattern the factory expects (`add_X`/`update_X`/`remove_X`) —
"list" and "get one" don't fit that create/update/remove shape, so they
were never candidates for the generic factory to begin with.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException

from api.deps import get_candidate_service
from app.candidate_service import CandidateService
from domain.models import CVDraft, CVDraftSummary

router = APIRouter(prefix="/candidates/{candidate_id}/drafts", tags=["cv-drafts"])


@router.get("")
def list_cv_drafts(service: CandidateService = Depends(get_candidate_service)) -> list[CVDraftSummary]:
    return service.list_cv_drafts()


@router.get("/{draft_id}")
def get_cv_draft(draft_id: str, service: CandidateService = Depends(get_candidate_service)) -> CVDraft:
    draft = service.get_cv_draft(draft_id)
    if draft is None:
        raise HTTPException(status_code=404, detail=f"CV draft {draft_id!r} not found.")
    return draft
